lib/stdlib/src/gen_server.erl

```diff
@@ -2292,27 +2305,39 @@ update_callback_cache(#server_data{module = Mod} = ServerData) ->
       handle_continue = fun Mod:handle_continue/2}.
 
 -doc false.
-wake_hib(ServerData, State, Debug) ->
+wake_hib(ServerData, State, Timer, Debug) ->
     Msg = receive
               Input ->
                   Input
           end,
-    decode_msg(update_callback_cache(ServerData), State, Msg, hibernate, Debug, true).
+    decode_msg(update_callback_cache(ServerData), State, Msg, Debug, Timer, true).
 
-decode_msg(#server_data{parent = Parent} = ServerData, State, Msg, Time, Debug, Hib) ->
+decode_msg(#server_data{parent = Parent, tag = Tag} = ServerData, State, Msg, Debug, Timer, Hib) ->
     case Msg of
         {system, From, Req} ->
             sys:handle_system_msg(Req, From, Parent, ?MODULE, Debug,
-                                  [ServerData, State, Time], Hib);
+                                  [ServerData, State, Timer, Hib], Hib);
         {'EXIT', Parent, Reason} ->
             terminate(ServerData, State, Msg, undefined, Reason, ?STACKTRACE(), Debug);
-        _Msg when Debug =:= [] ->
-            handle_msg(ServerData, State, Msg);
-        _Msg ->
-            Debug1 = sys:handle_debug(Debug, fun print_event/3,
-                                      ServerData#server_data.name, {in, Msg}),
-            handle_msg(ServerData, State, Msg, Debug1)
+	{timeout, TRef, Tag} when TRef =:= hd(Timer) ->
+            decode_msg(ServerData, State, tl(Timer), Debug);
+	{timeout, _Stale, Tag} ->
+	    LoopAction = case Hib of
+			     true -> hibernate;
+			     false -> infinity
+			 end,
+	    loop(ServerData, State, LoopAction, Debug, Timer);
```

Comment:
Maybe make this a new function `system_continue(ServerData, State, Hib, Debug, Timer)`  split out from `system_continue/3` to show that this goes the same way as when handling (ignoring) a system message.  Or it can be a function `loop_continue(ServerData, State, Hib, Debug, Timer)` that is also called from `system_continue/3`.

Reply:
@RaimoNiskanen done, I went for `loop_continue` :)